src/net/error_capturing_stream.rs

```diff
@@ -0,0 +1,130 @@
+use std::io::IoSlice;
+use std::net::SocketAddr;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use tokio::io::{self, AsyncRead, AsyncWrite, ReadBuf};
+
+use pin_project::pin_project;
+
+use crate::net::SessionStream;
+
+/// Stream that remembers the first error
+/// and keeps returning it afterwards.
+///
+/// It is needed to avoid accidentally using
+/// the stream after read timeout.
+#[derive(Debug)]
+#[pin_project]
+pub(crate) struct ErrorCapturingStream<T: AsyncRead + AsyncWrite + std::fmt::Debug> {
+    #[pin]
+    inner: T,
+
+    /// If true, the stream has already returned an error once.
+    ///
+    /// All read and write operations return error in this case.
+    is_broken: bool,
+}
```

Comment:
Just a possibility, you can decide whether to change something:

It would be nicer to make this an `enum { Ok(#[pin] T), Broken }`, but `get_ref()` and `get_ref_mut()` would need to be changed to return an Option. The callers of these functions would be fine, because `peer_addr()` is anyways called at creation time (since https://github.com/chatmail/core/pull/7017), and `set_read_timeout()` doesn't work anyways on a broken stream.